src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -82,8 +82,23 @@ public override void MoveFile(string sourceFullPath, string destFullPath)
         {
             // The desired behavior for Move(source, dest) is to not overwrite the destination file
             // if it exists. Since rename(source, dest) will replace the file at 'dest' if it exists,
-            // link/unlink are used instead. Note that the Unix FileSystemWatcher will treat a Move 
-            // as a Creation and Deletion instead of a Rename and thus differ from Windows.
+            // link/unlink are used instead. However, if the source path and the dest path refer to
+            // the same file, then do a rename rather than a link and an unlink.  This is important
+            // for case-insensitive file systems, so that we support changing the casing in the naming
+            // of the file. If this fails in any way (e.g. source file doesn't exist, dest file doesn't
+            // exist, rename fails, etc.), we just fall back to trying the link/unlink approach and
+            // generating any exceptional messages from there as necessary.
+            Interop.Sys.FileStatus sourceStat, destStat;
+            if (Interop.Sys.LStat(sourceFullPath, out sourceStat) == 0 && // source file exists
+                Interop.Sys.Stat(destFullPath, out destStat) == 0 && // dest file exists
```

Comment:
> you think this should be LStat instead?

I'm unsure. If `A` links to `B` and you try to rename `A` to `a` it won't give you the same file descriptor, right? But we should still be renaming, no?